core/src/p2p.rs

```diff
@@ -0,0 +1,741 @@
+        }
+    }
+
+    #[allow(
+        clippy::expect_used,
+        clippy::unwrap_in_result,
+        clippy::integer_division
+    )]
+    fn establish_new_connection(&self, listener: &TcpListener) -> Result<TcpStream, Error> {
+        let maybe_incoming_connection = {
+            let mut incoming_connection = None;
+            // ATTENTION!  This function is weird, and for good
+            // reason. We want to inject some stochastic sleeps as a
+            // form of Markov-chain-Monte-Carlo scheduling. This is
+            // necessary to avoid high contention and an actor-like
+            // model. We need to run the connection check randomly 7/8
+            // times.
+            //
+            // The precise timings needed for optimal performance are
+            // yet to be determined.
+            if rand::random::<u32>() % 8 > 1 {
+                for _ in 0_i32..20_i32 {
+                    if let Ok((stream, addr)) = listener.accept() {
+                        trace!(from=%addr, "Incoming p2p connection");
+                        stream
+                            .set_read_timeout(Some(P2P_TCP_TIMEOUT))
+                            .expect("Could not set read timeout on socket.");
+                        stream
+                            .set_write_timeout(Some(P2P_TCP_TIMEOUT))
+                            .expect("Could not set write timeout on socket.");
+                        // DO NOT SHORT-CIRCUIT to a return, this is
+                        // intentional for the timings.
+                        incoming_connection = Some(stream);
+                        break;
+                    }
+                    std::thread::sleep(ESTABLISH_CONNECTION_TIME_SLICE);
+                    if rand::random::<bool>() {
+                        std::thread::sleep(ESTABLISH_CONNECTION_TIME_SLICE * 2);
+                    }
+                }
+            }
+            incoming_connection
+        };
+
+        match maybe_incoming_connection {
+            Some(con) => Ok(con),
+            None => self
+                .try_resolve_address()
+                .map_or(Err(Error::Resolve), |addr| {
+                    TcpStream::connect_timeout(&addr, ESTABLISH_CONNECTION_TIME_SLICE * 40)
+                        .map(|stream| {
+                            info!("Outgoing p2p connection to {}", &addr);
+                            stream
+                                .set_read_timeout(Some(P2P_TCP_TIMEOUT))
+                                .expect("Could not set read timeout on socket.");
+                            stream
+                                .set_write_timeout(Some(P2P_TCP_TIMEOUT))
+                                .expect("Could not set write timeout on socket.");
+                            stream
+                        })
+                        .map_err(Error::Io)
+                }),
+        }
+    }
+
+    #[allow(clippy::unwrap_in_result, clippy::expect_used)]
```

Comment:
Seems to be redundant.

Reply:
It is. I'm going to work a little on `tracing` spans and `tracing_error` when the refactor is in an OK state.